src/planner/optimizer/secondary_index_scan/filter_expression_push_down.cpp

```diff
@@ -0,0 +1,663 @@
+        switch (expression->type()) {
+            case ExpressionType::kFunction: {
+                auto function_expression = std::static_pointer_cast<FunctionExpression>(expression);
+                auto const &f_name = function_expression->ScalarFunctionName();
+                static constexpr std::array<const char *, 5> IndexScanSupportedCompareFunctionNames = {"<", ">", "<=", ">=", "="};
+                // depth 0: <, >, <=, >=, = function
+                if (std::find(IndexScanSupportedCompareFunctionNames.begin(), IndexScanSupportedCompareFunctionNames.end(), f_name) !=
+                    IndexScanSupportedCompareFunctionNames.end()) {
+                    // supported compare function
+                    if (expression->arguments().size() != 2) {
+                        UnrecoverableError(fmt::format("Expression depth: {}. Unsupported expression type: In CheckExprIndexState(), the compare "
+                                                       "expression {} argument size != 2.",
+                                                       sub_expr_depth,
+                                                       expression->Name()));
+                        return false;
+                    }
+                    bool valid_compare = true;
+                    // left expression should be column
+                    auto is_column_index = [&m = candidate_column_index_map_](const SharedPtr<BaseExpression> &expr, u32 depth) -> bool {
+                        if (!(expr->Type().CanBuildSecondaryIndex())) {
+                            // Unsupported type
+                            LOG_TRACE(fmt::format("Expression depth: {}. In is_column_index(), unsupported column value type {}. Expression: {}.",
+                                                  depth,
+                                                  expr->Type().ToString(),
+                                                  expr->Name()));
+                            return false;
+                        }
+                        auto column_expression = std::static_pointer_cast<ColumnExpression>(expr);
+                        auto column_id = column_expression->binding().column_idx;
+                        if (m.contains(column_id)) {
+                            LOG_TRACE(fmt::format("Expression depth: {}. Column {} has index.", depth, expr->Name()));
+                            return true;
+                        } else {
+                            LOG_TRACE(fmt::format("Expression depth: {}. Column {} does not have a secondary index. Cannot apply index scan.",
+                                                  depth,
+                                                  expr->Name()));
+                            return false;
+                        }
+                    };
+                    if (!IsValidColumnExpression(expression->arguments()[0], sub_expr_depth + 1, is_column_index)) {
```

Comment:
left is column and right is value.
We also need to check:
right is value and left is column.